### DISCUSSION TITLE
Edge type that avoids collision with nodes

### DISCUSSION THREAD

**Discussion by jackparmer** [xyflow/xyflow#2806] [Ideas]
### Please describe the feature that you want to propose

For feedback loops like the below, it might be convenient to have an out-of-the-box edge type that calculates its path based on avoiding node positions

![](https://user-images.githubusercontent.com/53545754/215900187-3ecbb184-a8e6-4c01-9920-a70464a94a61.png)

Incredible library BTW :)

**Comment by jackparmer** [comment]
I suppose that one way that this could be done is to detect edge-node collisions upon connecting a node:
https://reactflow.dev/docs/examples/interaction/validation/

Then use a custom edge that calculates its trajectory to avoid node positions:
https://reactflow.dev/docs/examples/edges/custom-connectionline/

Seems like a lot of work though 😑 Other suggestions are very welcome.

Here's another example of where it would be nice to not have any edge and node overlap:

![image](https://user-images.githubusercontent.com/1865834/216684359-def251cf-e6c1-4a36-ba70-33365098ae24.png)

**Comment by moklick** [comment]
Hey @jackparmer !

We will probably not provide an edge algorithm in the near future (maybe never). There are too many different use cases and custom requirements. 

Possible solutions:

1. You could use a library like [Elkjs](https://github.com/kieler/elkjs) that comes with node and edge routing
2. You could implement a custom solution. For example an orthogonal edge routing algorithm [described in this article](https://medium.com/swlh/routing-orthogonal-diagram-connectors-in-javascript-191dc2c5ff70) 

Related / helpful package: https://github.com/tisoap/react-flow-smart-edge

**Comment by AndreRLima** [comment]
Hi @jackparmer,

Can I ask if you were able to find  a solution for this?

I tried [](https://github.com/tisoap/react-flow-smart-edge) but I could not achieve completely the behavior that I want

**Comment by Borouch** [comment]
Add me to the list of people who tried but not succeeded in this. 

If anyone has found solution would appreciate sharing it

**Reply by Riley1101** [reply]
Add me to the list of people who tried but not succeeded in this.

**Comment by awaisshah228** [comment]
I built an open-source package that solves this: **[avoid-nodes-edge](https://www.npmjs.com/package/avoid-nodes-edge)**

Unlike pathfinding-based approaches (like smart-edge which uses A* on a grid), this uses **libavoid** — a C++ routing engine compiled to WASM — producing clean orthogonal paths with automatic node avoidance, parallel edge nudging, configurable spacing/rounding, all running in a **Web Worker**.

**[Live Demo](https://avoid-nodes-pro-example.vercel.app)** | **[npm](https://www.npmjs.com/package/avoid-nodes-edge)** | **[GitHub](https://github.com/awaisshah228/react-flow-avoid-nodes-routing)** | **[Package source](https://github.com/awaisshah228/react-flow-avoid-nodes-routing/tree/turbo-package)**

```
npm install avoid-nodes-edge libavoid-js
```

Works with React Flow v12+.

**Comment by tisoap** [comment]
I'm the creator of [react-flow-smart-edge](https://github.com/tisoap/react-flow-smart-edge) , a library that uses using A* pathfinding to route edges around other nodes, avoiding intersections. It has smart equivalents for every type of edge React Flow has. It's been updated to work with React Flow 12.
